### ISSUE TITLE
Does CuPy work with PyPy?

### ISSUE THREAD

**Issue by leofang** [conda-forge/cupy-feedstock#43]
@conda-forge/cupy With #42, it seems our build matrix has 16 variants: CUDA (9.2, 10.0, 10.1, 10.2) x Python (3.6, 3.7, 3.8, PyPy 3.6). But does CuPy work with PyPy? I thought PyPy doesn't play nicely with Cython modules?

**Comment by jakirkham** [comment]
One of the PyPy developers has been doing work on this. AIUI this Cython and PyPy are compatible these days.

**Comment by leofang** [comment]
Looks like it's not working...
```
$ conda create -n pypy_cupy -c conda-forge pypy python cupy cudatoolkit=10.0
Collecting package metadata (current_repodata.json): done
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.7.12
  latest version: 4.8.3

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/leofang/miniconda3/envs/pypy_cupy

  added / updated specs:
    - cudatoolkit=10.0
    - cupy
    - pypy
    - python


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    bzip2-1.0.8                |       h516909a_2         396 KB  conda-forge
    certifi-2019.11.28         |   py36hc560c46_1         148 KB  conda-forge
    cupy-7.2.0                 |   py36h6e251b3_3        14.9 MB  conda-forge
    expat-2.2.9                |       he1b5a44_2         191 KB  conda-forge
    fastrlock-0.4              |py36h90709c0_1001          29 KB  conda-forge
    gdbm-1.18                  |       h0a1914f_1         192 KB  conda-forge
    libblas-3.8.0              |      16_openblas          10 KB  conda-forge
    libcblas-3.8.0             |      16_openblas          10 KB  conda-forge
    libffi-3.2.1               |    he1b5a44_1007          47 KB  conda-forge
    liblapack-3.8.0            |      16_openblas          10 KB  conda-forge
    libopenblas-0.3.9          |       h5ec1e0e_0         7.8 MB  conda-forge
    nccl-2.5.7.1               |       hd6f8bf8_0        96.6 MB  conda-forge
    numpy-1.18.1               |   py36he0f5f23_1         5.1 MB  conda-forge
    pypy-7.3.0                 |       h9f0ad1d_0           4 KB  conda-forge
    pypy3.6-7.3.0              |       h3e02ecb_7        31.1 MB  conda-forge
    python-3.6.9               |        0_73_pypy           4 KB  conda-forge
    python_abi-3.6             |    1_pypy36_pp73           4 KB  conda-forge
    setuptools-46.1.1          |   py36hc560c46_0         646 KB  conda-forge
    xz-5.2.4                   |    h516909a_1002         375 KB  conda-forge
    ------------------------------------------------------------
                                           Total:       157.5 MB

The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_llvm
  bzip2              conda-forge/linux-64::bzip2-1.0.8-h516909a_2
  ca-certificates    conda-forge/linux-64::ca-certificates-2019.11.28-hecc5488_0
  certifi            conda-forge/linux-64::certifi-2019.11.28-py36hc560c46_1
  cudatoolkit        pkgs/main/linux-64::cudatoolkit-10.0.130-0
  cudnn              pkgs/main/linux-64::cudnn-7.6.5-cuda10.0_0
  cupy               conda-forge/linux-64::cupy-7.2.0-py36h6e251b3_3
  expat              conda-forge/linux-64::expat-2.2.9-he1b5a44_2
  fastrlock          conda-forge/linux-64::fastrlock-0.4-py36h90709c0_1001
  gdbm               conda-forge/linux-64::gdbm-1.18-h0a1914f_1
  libblas            conda-forge/linux-64::libblas-3.8.0-16_openblas
  libcblas           conda-forge/linux-64::libcblas-3.8.0-16_openblas
  libffi             conda-forge/linux-64::libffi-3.2.1-he1b5a44_1007
  libgcc-ng          conda-forge/linux-64::libgcc-ng-9.2.0-h24d8f2e_2
  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-7.3.0-hdf63c60_5
  liblapack          conda-forge/linux-64::liblapack-3.8.0-16_openblas
  libopenblas        conda-forge/linux-64::libopenblas-0.3.9-h5ec1e0e_0
  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-9.2.0-hdf63c60_2
  llvm-openmp        conda-forge/linux-64::llvm-openmp-9.0.1-hc9558a2_2
  nccl               conda-forge/linux-64::nccl-2.5.7.1-hd6f8bf8_0
  ncurses            conda-forge/linux-64::ncurses-6.1-hf484d3e_1002
  numpy              conda-forge/linux-64::numpy-1.18.1-py36he0f5f23_1
  openssl            conda-forge/linux-64::openssl-1.1.1e-h516909a_0
  pip                conda-forge/noarch::pip-20.0.2-py_2
  pypy               conda-forge/noarch::pypy-7.3.0-h9f0ad1d_0
  pypy3.6            conda-forge/linux-64::pypy3.6-7.3.0-h3e02ecb_7
  python             conda-forge/linux-64::python-3.6.9-0_73_pypy
  python_abi         conda-forge/linux-64::python_abi-3.6-1_pypy36_pp73
  readline           conda-forge/linux-64::readline-8.0-hf8c457e_0
  setuptools         conda-forge/linux-64::setuptools-46.1.1-py36hc560c46_0
  six                conda-forge/noarch::six-1.14.0-py_1
  sqlite             conda-forge/linux-64::sqlite-3.30.1-hcee41ef_0
  tk                 conda-forge/linux-64::tk-8.6.10-hed695b0_0
  wheel              conda-forge/noarch::wheel-0.34.2-py_1
  xz                 conda-forge/linux-64::xz-5.2.4-h516909a_1002
  zlib               conda-forge/linux-64::zlib-1.2.11-h516909a_1006


Proceed ([y]/n)? y


Downloading and Extracting Packages
fastrlock-0.4        | 29 KB     | ############################################################################################################################ | 100% 
certifi-2019.11.28   | 148 KB    | ############################################################################################################################ | 100% 
pypy-7.3.0           | 4 KB      | ############################################################################################################################ | 100% 
libcblas-3.8.0       | 10 KB     | ############################################################################################################################ | 100% 
python_abi-3.6       | 4 KB      | ############################################################################################################################ | 100% 
nccl-2.5.7.1         | 96.6 MB   | ############################################################################################################################ | 100% 
numpy-1.18.1         | 5.1 MB    | ############################################################################################################################ | 100% 
libffi-3.2.1         | 47 KB     | ############################################################################################################################ | 100% 
setuptools-46.1.1    | 646 KB    | ############################################################################################################################ | 100% 
bzip2-1.0.8          | 396 KB    | ############################################################################################################################ | 100% 
libopenblas-0.3.9    | 7.8 MB    | ############################################################################################################################ | 100% 
pypy3.6-7.3.0        | 31.1 MB   | ############################################################################################################################ | 100% 
libblas-3.8.0        | 10 KB     | ############################################################################################################################ | 100% 
liblapack-3.8.0      | 10 KB     | ############################################################################################################################ | 100% 
gdbm-1.18            | 192 KB    | ############################################################################################################################ | 100% 
expat-2.2.9          | 191 KB    | ############################################################################################################################ | 100% 
cupy-7.2.0           | 14.9 MB   | ############################################################################################################################ | 100% 
python-3.6.9         | 4 KB      | ############################################################################################################################ | 100% 
xz-5.2.4             | 375 KB    | ############################################################################################################################ | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate pypy_cupy
#
# To deactivate an active environment, use
#
#     $ conda deactivate

$ conda activate pypy_cupy
$ python
Python 3.6.9 (1608da62bfc71e8ac775121dd0b21bb72e61c6ea, Mar 11 2020, 21:45:38)
[PyPy 7.3.0 with GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>>> import cupy
Traceback (most recent call last):
  File "/home/leofang/cupy/cupy/__init__.py", line 19, in <module>
    from cupy import core  # NOQA
  File "/home/leofang/cupy/cupy/core/__init__.py", line 1, in <module>
    from cupy.core import core  # NOQA
ImportError: cannot import name 'core'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/leofang/cupy/cupy/__init__.py", line 40, in <module>
    raise ImportError(msg) from e
ImportError: CuPy is not correctly installed.

If you are using wheel distribution (cupy-cudaXX), make sure that the version of CuPy you installed matches with the version of CUDA on your host.
Also, confirm that only one CuPy package is installed:
  $ pip freeze

If you are building CuPy from source, please check your environment, uninstall CuPy and reinstall it with:
  $ pip install cupy --no-cache-dir -vvvv

Check the Installation Guide for details:
  https://docs-cupy.chainer.org/en/latest/install.html

original error: cannot import name 'core'
```

**Comment by jakirkham** [comment]
Hmm... maybe we should pull the packages then and skip PyPy in the recipe? Looks like the sort of thing where we would need help from upstream for testing support. WDYT?

**Comment by leofang** [comment]
Yeah let me remove pypy in #44 first.

**Comment by leofang** [comment]
Tried building CuPy v7.3.0 locally with PyPy from conda-forge (but with cudatoolkit and friends removed; I used local CUDA instead). Can't reproduce what's seen above --- The build succeeded, and *CuPy was importable*, but almost all tests failed locally with similar errors:
```
cupy/cuda/compiler.py:287: in compile_with_cache
    extra_source, backend)
cupy/cuda/compiler.py:341: in _compile_with_cache_cuda
    ls.add_ptr_data(ptx, 'cupy.ptx')
cupy/cuda/function.pyx:230: in cupy.cuda.function.LinkState.add_ptr_data
    cpdef add_ptr_data(self, unicode data, unicode name):
cupy/cuda/function.pyx:232: in cupy.cuda.function.LinkState.add_ptr_data
    driver.linkAddData(self.ptr, driver.CU_JIT_INPUT_PTX, data_byte, name)
cupy/cuda/driver.pyx:198: in cupy.cuda.driver.linkAddData
    check_status(status)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   raise CUDADriverError(status)
E   cupy.cuda.driver.CUDADriverError: CUDA_ERROR_INVALID_IMAGE: device kernel image is invalid
```
I'm hoping it's something easy to fix, like `bytes` vs `str`...

**Comment by jakirkham** [comment]
That's an interesting data point. Thanks Leo! 😄

@isuruf, are there any catches when using compilers with PyPy that we might be missing here?

**Comment by isuruf** [comment]
`import cupy` works for me. Do you have `PYTHONPATH` or `PYTHONHOME` set?

**Comment by leofang** [comment]
No.

**Comment by isuruf** [comment]
`cupy` with CPython works for you? (Using conda)

Did you start python from a `cupy` source directory?

**Comment by leofang** [comment]
> `cupy` with CPython works for you? (Using conda)

Yes.

> Did you start python from a `cupy` source directory?

Ah, good catch, Isuru! Thanks. I think this is it...Sorry for being stupid 😞 I can import condo-forge cupy in pypy now. But actually running it would encounter the error seen in https://github.com/conda-forge/cupy-feedstock/issues/43#issuecomment-603032749. I think this is a CuPy issue, though, not packaging.

**Comment by leofang** [comment]
Note to self: PyPy build is disabled starting #44.

**Comment by jakirkham** [comment]
Raised issue ( https://github.com/cupy/cupy/issues/3221 ) upstream about PyPy support.

**Comment by leofang** [comment]
Ah, thanks John! Having limited bandwidth today...

**Comment by leofang** [comment]
It seems @okuta fixed it in #3225 and it will be backported to v7 shortly. I will test it locally shortly. I think we can bring back the pypy build in v7.4.0 and close this issue.

**Comment by jakirkham** [comment]
SGTM. Glad it was straightforward to fix 😀
